Import services, interactions and service interactions

davidbasalla opened the pull request:
This PR adds importer classes and rake tasks to import Services, Interactions and ServiceInteractions from standards.esd website. The original files are in CSV format, so we download them with Net:HTTP's `get_response` method, parse them and store them as records in our local database. The order of the imports is important, as Services and Interactions have to be imported first, before ServiceInteractions (which acts as a join model between the two) can be imported successfully.

We split out a `CsvDownloader` class which we use in our importer classes as well as `LocalAuthorityUrlImporter`.

For ServiceInteractions we log a summary of imported rows, as there are a number of errors that can appear.

Worked with @klssmith and @emmabeynon on this story, just plastered my face on all commits as we were using my laptop :)

davidbasalla committed:
Update LA URL Importer to use CsvDownloader

Having created a custom CsvDownloader class to handle CSV importing, we refactor `LocalAuthoritiesUrlImporter` and its tests to make use of it.

davidbasalla committed:
Create CsvDownloader class

We create a class for downloading CSV files which we can reuse in other classes to import CSV data. We use the Net::HTTP module and match against Net::HTTPOK for success. We raise a custom error if there is no match, partially in order to get around having to rescue the large number of Net::HTTP errors

davidbasalla committed:
Create ServiceImporter class

We create the class to import Services from a preset URL. We use the `CsvDownloader` class to import the CSV data, using Dependency Injection (for easier testing). We log errors if the import fails. For each imported CSV row, the importer either creates a new record or updates an existing record.

davidbasalla committed:
Create InteractionsImporter class

We create the class to import Interactions from a preset URL. We use the `CsvDownloader` class to import the CSV data, using Dependency Injection (for easier testing). We log errors if the import fails. For each imported CSV row, the importer either creates a new record or updates an existing record.

davidbasalla committed:
Create ServiceInteractionsImporter class

We create the class to import ServiceInteractions from a preset URL. We use the `CsvDownloader` class to import the CSV data, using Dependency Injection (for easier testing). After the import, we check that the associated Service and Interaction have already been created, raising errors if not.
We log errors if the import fails. For each imported CSV row, the importer either creates a new record or updates an existing record.

davidbasalla committed:
Create import task for ServiceInteractions

The rake task groups all import tasks for Services, Interactions and ServiceInteractions. The order is important so that Services and Interactions are fully imported before ServiceInteractions which links them together. Therefore, the rake task should be invoked with `import_all`

h-lame commented:
This all looks good to me.  Although the description of the pr still talks about `open_uri` whereas the implementation now uses `Net::HTTP`.

davidbasalla committed:
Add detailed summary logging for ServiceInteractions

When importing ServiceInteractions, we have identified a number of things that can wrong, such as missing fields in the CSV or linking to Services or Interactions that that do not exist.
Previously we only log when something goes wrong for each entry, which is hard to read in the logs. This commit adds a convenient summary into the log at the end of the import.

h-lame commented on lib/local-links-manager/import/service_interactions_importer.rb:
```diff
@@ -70,6 +81,14 @@ def create_or_update_record(parsed_hash)
+      def import_summary
+        "ServiceInteraction Import complete\n"\
+        "Downloaded CSV rows: #{downloaded_csv_rows.count}\n"\
+        "Imported records: #{ServiceInteraction.count}\n"\
```
This isn't the number of imported records, it's the number of service interactions in the system.  I'm splitting hairs a bit, but we could have 10 rows in the CSV and 100 rows already imported, which is going to make this number look weird.  What we probably want here is `downloaded_csv_rows.count - missing_id_count - missing_record_count` - or we keep track of the successful rows like we do for the missing id and missing record ones.

davidbasalla replied:
Thank you, have addressed your comment. I went with another count variable for successful rows added, mainly because I'm not 100% sure that the code catches all errors (there might be more than missing ids and missing associated records)

issyl0 commented:
Seems fine. ⭐
